[feat] Sandbox: support sandbox fusion on FaaS & localhost

chenhaiq opened the pull request:
### Checklist Before Starting

- [ *] Search for similar PR(s).

### What does this PR do?

Implement sandbox fusion backend on FaaS. For example, reward score using a FaaS instance on volcengine.com. It have better performance and security comparing to local sandbox.

### Specific Changes

Added a code branch in _default_compute_score to choose sandbox according to sandbox_fusion_url configuration. 


### Usage Example

examples/ppo_trainer/run_deepseek7b_llm_sandbox_fusion.sh

### Test

tests/reward_score/test_sandbox_fusion.py
However, the new testcase requires to setting Sandbox API URL in env SANDBOX_FUSION_URL. If the env is not set, most testcases will be skipped.

### Additional Info.

Using sandbox on Faas have save 60% time on reward process comparing local sandbox:
<img width="273" alt="截屏2025-05-07 20 37 05" src="https://github.com/user-attachments/assets/fc9c0e23-6afe-4f34-a28a-a1756e85d45f" />


### Checklist Before Submitting

- [*] Read the [Contribute Guide](https://github.com/volcengine/verl?tab=readme-ov-file#contribution-guide).
- [*] Apply [pre-commit checks](https://github.com/volcengine/verl?tab=readme-ov-file#code-linting-and-formatting).
- [*] Add `[BREAKING]` to the PR title if it breaks any API.
- [*] Update the documentation about your changes in the [docs](https://github.com/volcengine/verl/tree/main/docs).
- [*] Add CI test(s) if neccessary.

wuxibin89 commented on verl/utils/reward_score/sandbox_fusion/utils.py:
For multi turn rollout, each chat completion may independently call check_correctness, and we may have thousands of chat completion running concurrently. So we better have a global ThreadPoolExecutor, or more ideally use asyncio to avoid invoke hundreds of thread.

A `SandboxFusionManager` may a better option for  thread pool/event loop, rate limit control, etc.
```python
class SandboxFusionManager:
    def submit(self, in_outs: Optional[dict], generation: str, ...):
        pass
```

chenhaiq replied:
1. implemented a global concurrent limit to fusion api calls by using a thread semaphore
2. since prime.py calls compute_score method in subprocess concurrently, it can not work with asyncio in sandbox fusion. I will submit another version of sandbox for asyncio soon with SandboxFusionManager.

eric-haibin-lin commented on verl/trainer/config/ppo_megatron_trainer.yaml:
```diff
@@ -229,7 +229,8 @@ reward_model:
   use_dynamic_bsz: ${critic.use_dynamic_bsz}
   max_length: null
   launch_reward_fn_async: False # custom reward function executed async on CPU, during log_prob
-
```
let's make it hierarchical:
sandbox_fusion:
   url:
   xxxx:

chenhaiq replied:
Pushed a new commit to fix this. Please review again.

eric-haibin-lin commented on tests/reward_score/test_sandbox_fusion.py:
```diff
@@ -0,0 +1,540 @@
+
+@patch("verl.utils.reward_score.sandbox_fusion.utils.call_sandbox_api")
+def test_unit_concurrency_order(mock_call_sandbox_api):
+    """单元测试：验证并发执行时结果顺序是否正确"""
```
please leave docs in English, thank you. Same for L238, L244, L255, L257, L265, and a few more occurrences

chenhaiq replied:
> please leave docs in English, thank you. Same for L238, L244, L255, L257, L265, and a few more occurrences

fixed

eric-haibin-lin commented on verl/utils/reward_score/sandbox_fusion/__init__.py:
```diff
@@ -0,0 +1,95 @@
+
+"""
+Verify code correctness using the sandbox_fusion FaaS service provided by public clouds. Public clouds covered by testing include:
+- volcengine.com
```
is it possible to deploy sandbox fusion locally, and does this PR support that kind of usage as well? How do users develop and debug sandbox before onboarding with volc-engine?

chenhaiq replied:
update this comment

chenhaiq commented:
@eric-haibin-lin ：
1. added a document about how to use sandbox fusion
2. fixed a performance issue because reasoning LLM may generate very long context without code.

eric-haibin-lin commented on verl/utils/reward_score/prime_code/__init__.py:
```diff
@@ -50,7 +50,7 @@ def compute_score(completion, test_cases, continuous=False):
             res_list = []
             for test_case_id, test_case in enumerate(test_cases_list):
-                res, metadata = apps_check_correctness(in_outs=test_case, generation=solution, timeout=5, debug=False)
+                res, metadata = apps_check_correctness(in_outs=test_case, generation=solution, timeout=10, debug=False)
```
why changing the timeout? What is the unit used here (second, minute?)

chenhaiq replied:
It is because the local sandbox changed timeout from 10 to 5. Using the same default value to have the same effect.

chenhaiq commented:
@eric-haibin-lin the test failed because HTTPSConnectionPool(host='hf-mirror.com', port=443): Read timed out. Can we rerun it?
